### PR TITLE
refactor(sql): align datatype conversions with the new convention

### PR DESCRIPTION
**Pull request by kszucs**
Updated all the SQL backends to follow the new convention using:
- `dtype_from_<backend>`
- `dtype_to_<backend>`

### CODE REVIEW

File: ibis/backends/base/sql/alchemy/translator.py [hunk 1]
```diff
@@ -69,6 +74,9 @@ def integer_to_timestamp(self, arg, tz: str | None = None):
 
     supports_unnest_in_select = True
 
+    get_sqla_type = staticmethod(dtype_to_sqlalchemy)
```
Comment (kszucs) [inline ibis/backends/base/sql/alchemy/translator.py:77]: Perhaps we should figure out a nicer way to inject the converter functions to the backend classes, any idea?

File: ibis/backends/mssql/compiler.py [hunk 2]
```diff
@@ -27,6 +25,10 @@ class MsSqlExprTranslator(AlchemyExprTranslator):
     _require_order_by = AlchemyExprTranslator._require_order_by + (ops.Reduction,)
     _dialect_name = "mssql"
 
+    @classmethod
```
Comment (kszucs) [inline ibis/backends/mssql/compiler.py:28]: Use staticmethod here.

File: ibis/backends/snowflake/datatypes.py [hunk 3]
```diff
@@ -38,88 +41,55 @@
...
+        return dt.Timestamp(timezone=None, nullable=nullable)
+    elif isinstance(typ, (TIMESTAMP_LTZ, TIMESTAMP_TZ)):
+        return dt.Timestamp(timezone="UTC", nullable=nullable)
+    elif isinstance(typ, ARRAY):
+        return dt.Array(dt.json, nullable=nullable)
+    elif isinstance(typ, OBJECT):
+        return dt.Map(dt.string, dt.json, nullable=nullable)
+    elif isinstance(typ, VARIANT):
+        return dt.JSON(nullable=nullable)
+    elif isinstance(typ, sa.Numeric):
```
Comment (gforsyth) [inline ibis/backends/snowflake/datatypes.py:84]:
```suggestion
    elif isinstance(typ, sat.Numeric):
```

File: ibis/backends/duckdb/__init__.py [hunk 4]
```diff
@@ -114,7 +115,7 @@ def column_reflect(inspector, table, column_info):
                         _, typ, *_ = con.connection.fetchone()
                     complex_type_info_cache[colname] = coltype = parse(typ)
 
-                column_info["type"] = to_sqla_type(engine.dialect, coltype)
+                column_info["type"] = dtype_to_sqlalchemy(coltype)
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/duckdb/__init__.py:118]: It should, thanks for catching!

File: ibis/backends/base/sql/alchemy/datatypes.py [hunk 5]
```diff
@@ -134,102 +130,113 @@ class Unknown(sa.Text):
...
-    pass
-else:
+    Parameters
+    ----------
+    dtype :
+        Ibis type to convert.
+    converter : ]
+        Converter function to use for nested types. If not provided, this function
+        will be used recursively. Should only be used when defining new converter for
+        dialects.
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/base/sql/alchemy/datatypes.py:202]: Updated.

File: ibis/backends/base/sql/alchemy/datatypes.py [hunk 6]
```diff
@@ -238,157 +245,49 @@ def sa_uuid(_, satype, nullable=True):
...
-    for column in table.columns:
-        name = column.name
-        if name in schema:
-            dtype = dt.dtype(schema[name])
-        else:
-            dtype = dt.dtype(dialect, column.type, nullable=column.nullable)
-        pairs.append((name, dtype))
-    return sch.schema(pairs)
+def dtype_from_sqlalchemy(typ, nullable=True, converter=None):
+    convert = converter or dtype_from_sqlalchemy
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/base/sql/alchemy/datatypes.py:261]: Added.

File: ibis/backends/base/sql/alchemy/datatypes.py [hunk 7]
```diff
@@ -238,157 +245,49 @@ def sa_uuid(_, satype, nullable=True):
...
+
+    if dtype := _from_sqlalchemy_types.get(type(typ)):
+        return dtype(nullable=nullable)
+    elif isinstance(typ, sat.Float):
+        if (dectyp := _FLOAT_PREC_TO_TYPE.get(typ.precision)) is not None:
+            return dectyp(nullable=nullable)
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/base/sql/alchemy/datatypes.py:267]: Updated.

File: ibis/backends/duckdb/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/duckdb/__init__.py:118]: Well, I don't have access to the compiler here, so updated it to be `dtype_from_duckdb`

File: ibis/backends/snowflake/datatypes.py (hunk 3 above)
Comment (kszucs) [reply ibis/backends/snowflake/datatypes.py:84]: Updated.

File: ibis/backends/duckdb/datatypes.py [hunk 8]
```diff
@@ -1,13 +1,14 @@
 from __future__ import annotations
 
+import duckdb_engine.datatypes as duckt
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/duckdb/datatypes.py:3]: Quack!

File: ibis/backends/base/sql/alchemy/__init__.py [hunk 9]
```diff
@@ -478,6 +476,42 @@ def _get_sqla_table(
                 return table
             return self._handle_failed_column_type_inference(table, nulltype_cols)
 
+    # TODO(kszucs): remove the schema parameter
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/base/sql/alchemy/__init__.py:479]: Just a nice to have.

File: ibis/backends/snowflake/datatypes.py [hunk 10]
```diff
@@ -11,14 +9,19 @@
...
-from ibis.backends.base.sql.alchemy import to_sqla_type
+from ibis.backends.base.sql.alchemy.datatypes import (
+    dtype_from_sqlalchemy,
+    dtype_to_sqlalchemy,
+)
+
+
+@compiles(sat.NullType, "snowflake")
+def compiles_nulltype(element, compiler, **kw):
+    return "VARIANT"
```
Comment (cpcloud) [inline ibis/backends/snowflake/datatypes.py:23]: Is this necessary for some reason?

**Review by cpcloud — APPROVED** [review]
LGTM! 🚢 it!

**Comment by kszucs** [comment]
I forgot to update the commit message, so I forced pushed to the main branch with updated commit message, here is the commit for reference https://github.com/ibis-project/ibis/commit/0ef145b64a8ac8893ea52cb5bdf9b14952ffc91c

**Pull request merged**